### PR TITLE
Add generic support to Collection and it's methods

### PR DESCRIPTION
**Pull request by beary**
Please fill in this template.

- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [X] Provide a URL to documentation or source code which provides context for the suggested changes: <<http://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-3.html>>
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dslint/dt.json" }`.

TypeScript support default generic since version 2.3, So I add generic to collection. When create a collection by `DB.collection()`or`DB.createCollection()`, developer could specify a scheme, and when do some query, the result can be static analysd.

These changes don't break the old cold.

eg:
```typescript
import { MongoClient } from 'mongodb'

interface User {
  name: string
  gender: number
}

interface SomeType {
  // ...
}


MongoClient.connect('...').then(DB => {
  // It's also support old code
  DB
    .collection('User')
    .find()
    .toArray()
    .then(arr => {
      arr.forEach(user => {
        // user: any
      })
    })

  DB
    .collection<User>('User')
    .find()
    .toArray()
    .then(arr => {
      arr.forEach(user => {
        console.log(user.gender) // user: User
      })
    })

  DB
    .collection<User>('User')
    .aggregate<SomeType>([])
    .toArray()
    .then(arr => {
      arr.forEach(item => {
        // item: SomeType
      })
    })

  DB
    .collection<User>('User')
    .findOne({})
    .then(user => {
      // user: User
    })

  DB
    .collection<User>('User')
    .findOne<SomeType>({}, { fields: {/* ... */ } }) // If pass fields option, could specify generic to `findOne()`
    .then(user => {
      // user: SomeType
    })
})
```

### CODE REVIEW

**Comment by beary** [comment]
I checked logs, it says there are some syntax error. But there is no problem on my code, so strange... :cry: 
My versions:
typescript: 2.3.2
vscode: code-insiders_1.12.0-1493823801_amd64.deb

```shell
# test logs
../mongodb/index.d.ts(184,24): error TS1005: ',' expected.
../mongodb/index.d.ts(185,24): error TS1005: ',' expected.
../mongodb/index.d.ts(186,24): error TS1005: ',' expected.
../mongodb/index.d.ts(195,30): error TS1005: ',' expected.
../mongodb/index.d.ts(196,30): error TS1005: ',' expected.
../mongodb/index.d.ts(197,30): error TS1005: ',' expected.
../mongodb/index.d.ts(239,30): error TS1005: ',' expected.
../mongodb/index.d.ts(240,30): error TS1005: ',' expected.
../mongodb/index.d.ts(241,30): error TS1005: ',' expected.
../mongodb/index.d.ts(424,17): error TS1005: ',' expected.
../mongodb/index.d.ts(425,17): error TS1005: ',' expected.
../mongodb/index.d.ts(466,12): error TS1005: ',' expected.
../mongodb/index.d.ts(469,15): error TS1005: ',' expected.
../mongodb/index.d.ts(470,15): error TS1005: ',' expected.
```

**Comment by CaselIT** [comment]
@beary the build fails because of https://github.com/Microsoft/dtslint/issues/31. Shortly DT will start supporting features that require the latest version of TS about a month after its release. 

Otherwise LGTM 👍

**Comment by beary** [comment]
Thanks for your response @CaselIT

File: types/mongodb/index.d.ts [hunk 1]
```diff
@@ -463,26 +461,25 @@ export interface Collection {
     dropIndexes(): Promise<any>;
     dropIndexes(callback?: MongoCallback<any>): void;
     //http://mongodb.github.io/node-mongodb-native/2.1/api/Collection.html#find
-    find(query?: Object): Cursor<any>;
-    find<T>(query?: Object): Cursor<T>;
+    find(query?: Object): Cursor<TSchema>;
     /** @deprecated */
-    find(query: Object, fields?: Object, skip?: number, limit?: number, timeout?: number): Cursor<any>;
+    find<T = TSchema>(query: Object, fields?: Object, skip?: number, limit?: number, timeout?: number): Cursor<T>;
```
Comment (andy-ms) [inline types/mongodb/index.d.ts]: Is `T` ever *not* `TSchema`?

**Review by ghost — COMMENTED** [review]
Needs a `// TypeScript Version: 2.3` comment on line 5.

File: types/mongodb/index.d.ts (hunk 1 above)
Comment (beary) [inline types/mongodb/index.d.ts]: Yes, When `fields` parameter was passed in, the results would not have some properties which `TSchema` has. This is an old api [http://mongodb.github.io/node-mongodb-native/1.4/api-generated/collection.html#find](http://mongodb.github.io/node-mongodb-native/1.4/api-generated/collection.html#find )

Comment (beary) [inline types/mongodb/index.d.ts]: Sorry, my expression was puzzled. Actually I means `T` always same as `TSchema`, but if developer pass `fields` parameter, they can specify `T` according their `fields`

**Comment by yuit** [comment]
@beary  if you could add the `// TypeScript Version: 2.3` it should resolve your build failure and I can promptly merge the PR. Thanks for your contribution!

File: types/mongodb/index.d.ts [hunk 2]
```diff
@@ -2,10 +2,11 @@
 // Project: https://github.com/mongodb/node-mongodb-native/tree/2.2
 // Definitions by: Federico Caselli <https://github.com/CaselIT>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
-// TypeScript Version: 2.3
 
 // Documentation : http://mongodb.github.io/node-mongodb-native/2.2/api/
 
+// TypeScript Version: 2.3
```
Comment (andy-ms) [inline types/mongodb/index.d.ts]: Old location was correct.
Packages that depend on mongo will also have to be updated to have this in their header. #16198 is doing this.

**Comment by hinell** [comment]
@beary I had the same problem when have tried to migrate ``mongodb`` module to 2.3 ts version.
If you want to solve travis error you have to add ts 2.3 header to all these modules :neutral_face: :
```
The following packages had errors: acl, agenda, connect-mongo, easy-jsend, express-brute-mongo,
gridfs-stream, mongoose, mongoose-auto-increment, mongoose-deep-populate, mongoose-mock,
mongoose-paginate, mongoose-promise, mongoose-sequence, passport-local-mongoose
```

**Comment by CaselIT** [comment]
@beary @hinell @andy-ms I'll be doing the  same thing for this PR https://github.com/DefinitelyTyped/DefinitelyTyped/pull/16198 where it's still missing a couple of definitions. Maybe its best to merge this after the other one, to avoid duplicating the work?

**Comment by hinell** [comment]
@CaselIT Are you going to do it manually? If so It is pretty a lot of job there. I think it would be more better quicker if somebody would write script that automatically reset all headers in all specified packages.

**Comment by beary** [comment]
@hinell Thanks, @CaselIT add `// TypeScript Version 2.3` to all those modules on  https://github.com/DefinitelyTyped/DefinitelyTyped/pull/16198 , I'll waiting for his PR was merged.

@CaselIT No problem. I can do my work after your PR was merged.

**Comment by hinell** [comment]
@CaselIT At first I didn't understand you. See my post above again pls. 😄

**Comment by CaselIT** [comment]
@hinell Yes I did it manually. It's not too much work

File: types/mongodb/index.d.ts [hunk 3]
```diff
@@ -743,9 +741,9 @@ export interface DeleteWriteOpResultObject {
 }
 
 //http://mongodb.github.io/node-mongodb-native/2.1/api/Collection.html#~findAndModifyWriteOpResult
-export interface FindAndModifyWriteOpResultObject {
+export interface FindAndModifyWriteOpResultObject<TSchema> {
```
Comment (CaselIT) [inline types/mongodb/index.d.ts]: I think it's better to set `<TSchema=Default>`

**Review by CaselIT — COMMENTED** [review]
Apart from the note I've left it looks good. 👍

**Comment by beary** [comment]
@CaselIT Thanks for remind, default type argument for `FindAndModifyWriteOpResultObject` was added. I realize sometimes people would use the interface as a type annotations, the default type argument is necessary.

**Comment by beary** [comment]
@yuit Thank you!

**Pull request merged**